source/common/upstream/cluster_manager_impl.cc

```diff
@@ -207,15 +208,58 @@ ClusterManagerImpl::ClusterManagerImpl(const envoy::api::v2::Bootstrap& bootstra
   }
 
   for (const auto& cluster : bootstrap.static_resources().clusters()) {
-    loadCluster(cluster, false);
+    // First load all the primary clusters.
+    if (cluster.type() != envoy::api::v2::Cluster::EDS) {
+      loadCluster(cluster, false);
+    }
   }
 
-  // We can now potentially create the CDS API once the backing cluster exists.
-  if (bootstrap.dynamic_resources().has_cds_config()) {
-    cds_api_ = factory_.createCds(bootstrap.dynamic_resources().cds_config(), eds_config_, *this);
-    init_helper_.setCds(cds_api_.get());
-  } else {
-    init_helper_.setCds(nullptr);
+  for (const auto& cluster : bootstrap.static_resources().clusters()) {
+    // Now load all the secondary clusters.
```

Comment:
Can you add some explanation here on the distinction between primary and secondary clusters?

Reply:
The one edge case that remains here is that I am not being strict about EDS clusters backed by EDS clusters. That is, if when we load EDS cluster `foo`, we have already loaded EDS cluster `bar` (which `foo` depends on) we let it be. However, if we load `bar` then `foo`, we throw the exception.

I have toyed with 3 ways to reason this:

1. We prohibit EDS clusters backed by EDS clusters.
2. We allow EDS clusters backed by EDS clusters, at which point I bifurcate to:
  a.  We leave the change as is, and now that we have understandable exceptions people can reason about add order and can reorganize their statically loaded clusters to fix the problem.
  b. I make the code even more complex, where we have a third verification pass after all clusters have been added and verify that all EDS clusters backed by EDS clusters are valid.

Reply:
What is an EDS cluster backed by an EDS cluster?

Reply:
I see, you mean an EDS cluster that uses a config source that is another EDS cluster. We should block this I think, so (1).

Reply:
Yes that is exactly what I mean. Ok, sounds good. I'll update